lib/api/api-fetch/body.js

```diff
@@ -1,66 +1,183 @@
 
 // https://fetch.spec.whatwg.org/#concept-bodyinit-extract
 function extractBody (body) {
   // TODO: FormBody
 
+  let stream = null
+  let source = null
+  let length = null
+  let type = null
+
   if (body == null) {
     return [null, null]
   } else if (body instanceof URLSearchParams) {
     // spec says to run application/x-www-form-urlencoded on body.list
     // this is implemented in Node.js as apart of an URLSearchParams instance toString method
     // See: https://github.com/nodejs/node/blob/e46c680bf2b211bbd52cf959ca17ee98c7f657f5/lib/internal/url.js#L490
     // and https://github.com/nodejs/node/blob/e46c680bf2b211bbd52cf959ca17ee98c7f657f5/lib/internal/url.js#L1100
-    return [{
-      source: body.toString()
-    }, 'application/x-www-form-urlencoded;charset=UTF-8']
-  } else if (typeof body === 'string') {
-    return [{
-      source: body
-    }, 'text/plain;charset=UTF-8']
+    source = body.toString()
+    length = Buffer.byteLength(source)
+    type = 'application/x-www-form-urlencoded;charset=UTF-8'
   } else if (body instanceof ArrayBuffer || ArrayBuffer.isView(body)) {
-    return [{
-      source: body
-    }, null]
+    if (body instanceof DataView) {
+      // TODO: Blob doesn't seem to work with DataView?
+      body = body.buffer
+    }
+    source = body
+    length = Buffer.byteLength(body)
   } else if (util.isBlob(body)) {
-    return [{
-      source: body,
-      length: body.size
-    }, body.type || null]
-  } else if (util.isStream(body) || typeof body.pipeThrough === 'function') {
+    source = body
+    length = body.size
+    type = body.type
+  } else if (typeof body.pipeThrough === 'function') {
     if (util.isDisturbed(body)) {
       throw new TypeError('disturbed')
     }
 
-    let stream
-    if (util.isStream(body)) {
-      stream = Readable.toWeb(body)
-    } else {
-      if (body.locked) {
-        throw new TypeError('locked')
+    if (body.locked) {
+      throw new TypeError('locked')
+    }
+
+    stream = body
+  } else {
+    source = String(body)
+    length = Buffer.byteLength(source)
+    type = 'text/plain;charset=UTF-8'
+  }
+
+  if (!stream) {
+    assert(source != null)
+    if (!ReadableStream) {
+      ReadableStream = require('stream/web').ReadableStream
+    }
+    stream = new ReadableStream({
+      async start (controller) {
+        controller.enqueue(source)
+        controller.close()
+      },
+      async pull () {
+      },
+      async cancel (reason) {
+      }
+    })
+  }
+
+  return [{ stream, source, length }, type]
+}
+
+function cloneBody (src) {
+  if (!src) {
+    return
+  }
+
+  if (util.isDisturbed(src)) {
+    throw new TypeError('disturbed')
+  }
+
+  if (src.locked) {
+    throw new TypeError('locked')
+  }
+
+  // https://fetch.spec.whatwg.org/#concept-body-clone
+  const [out1, out2] = src.stream.tee()
+  src.stream = out1
+  return {
+    stream: out2,
+    length: src.length,
+    source: src.source
+  }
+}
+
+class Body {
```

Comment:
Not required for this PR, but Body isn't really a class per the spec. Its just a mixin and adds some things to the Request and Response class. This is slightly important to how those classes are implemented because if `Request` has to `extend Body` in order to get access to these methods and properties, then it must also call `super()` in the constructor and that breaks the specified `Request` and `Response` initialization steps. That is the main reason why i have Body implemented  as a function in undici-fetch (https://github.com/Ethan-Arrowood/undici-fetch/blob/f400f890d19034d00562763f472c0d2f5824bcb6/src/body.js#L27)

Reply:
Yea I didn't want it as a class but didn't manage to get a mixin working.

Reply:
Make an issue and feel free to open a PR once this has landed!

Reply:
I fixed it based on your link. Thanks!